### PR TITLE
plugin/forward: add new proxy implementation

### PR DESCRIPTION
**Pull request by miekg**
This is a new proxy implementation that caches upstream connections. It
works with udp and tcp. The healthchecks are inband doing a ". IN NS"
+norec query against the upstream.

See the README.md for the implemented features.

### CODE REVIEW

**Comment by miekg** [comment]
~~TODO: tests~~

**Comment by johnbelamaric** [comment]
This isn't coming out faster in my tests: https://docs.google.com/document/d/183DsEDkRn8ceM_QWNSrDotTmJVl1qoJrila0H0ol44U/edit?ts=59cea451#

File: plugin/forward/README.md [hunk 1]
```diff
@@ -79,7 +79,15 @@ Proxy everything except `example.org` using the host's `resolv.conf`'s nameserve
 ~~~ corefile
 . {
     proxy . /etc/resolv.conf {
```
Comment (johnbelamaric) [inline plugin/forward/README.md]: s/proxy/forward/

**Comment by johnbelamaric** [comment]
Crashes. :(

After a while, I got:

```
.:53
2017/10/05 19:55:11 [INFO] CoreDNS-011
2017/10/05 19:55:11 [INFO] linux/amd64, go1.8.3, f23da05-dirty
CoreDNS-011
linux/amd64, go1.8.3, f23da05-dirty
fatal error: concurrent map read and map write

goroutine 70 [running]:
runtime.throw(0x182d469, 0x21)
	/usr/local/go/src/runtime/panic.go:596 +0x95 fp=0xc4204a9db8 sp=0xc4204a9d98
runtime.mapaccess2_faststr(0x162cf00, 0xc42037af90, 0xc43e441360, 0x16, 0x1806a34, 0x3)
	/usr/local/go/src/runtime/hashmap_fast.go:326 +0x50a fp=0xc4204a9e18 sp=0xc4204a9db8
github.com/coredns/coredns/plugin/forward.(*proxy).handleClientPackets(0xc4202af500)
	/go/src/github.com/coredns/coredns/plugin/forward/proxy.go:112 +0x194 fp=0xc4204a9fd8 sp=0xc4204a9e18
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2197 +0x1 fp=0xc4204a9fe0 sp=0xc4204a9fd8
created by github.com/coredns/coredns/plugin/forward.(*Forward).Start
	/go/src/github.com/coredns/coredns/plugin/forward/setup.go:50 +0x9b

goroutine 1 [semacquire]:
sync.runtime_Semacquire(0xc42038010c)
	/usr/local/go/src/runtime/sema.go:47 +0x34
sync.(*WaitGroup).Wait(0xc420380100)
	/usr/local/go/src/sync/waitgroup.go:131 +0x7a
github.com/mholt/caddy.(*Instance).Wait(0xc420299110)
	/go/src/github.com/mholt/caddy/caddy.go:412 +0x2f
github.com/coredns/coredns/coremain.Run()
	/go/src/github.com/coredns/coredns/coremain/run.go:105 +0x206
main.main()
	/go/src/github.com/coredns/coredns/coredns.go:8 +0x20

goroutine 18 [syscall]:
os/signal.signal_recv(0x0)
	/usr/local/go/src/runtime/sigqueue.go:116 +0x104
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:22 +0x22
created by os/signal.init.1
	/usr/local/go/src/os/signal/signal_unix.go:28 +0x41
```

336MB file, that's just the head of course.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] WIP: plugin/f..." ]
>Crashes. :(
>
>After a while, I got:

Think I found it. Tried to normalize these locks, but a lot of different things 
happen between the various locking.

**Comment by miekg** [comment]
found some more races (thanks tests! :) ). Those found all fixes of course.

**Comment by miekg** [comment]
Ready for review (and for discussion) if this should be kept external or not.

**Comment by miekg** [comment]
Closing here in favour for https://github.com/miekg/forward
I'll put this in the external plugins on coredns.io, we can take any inclusion discussion from there.

**Pull request closed without merging**